### PR TITLE
Flatbuffer for Plasma

### PR DESCRIPTION
**Pull request by pcmoritz**

### CODE REVIEW

File: src/plasma/plasma_protocol.c [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+    flatbuffers_string_vec_push(B, id);
+  }
+  return flatbuffers_string_vec_end(B);
+}
+
+/**
+ * Reads an array of object IDs from a flatbuffer vector.
+ *
+ * @param object_id_vector Flatbuffer vector containing object IDs.
+ * @param object_ids_ptr Pointer to array that will contain the object IDs. The
```
Comment (robertnishihara) [inline src/plasma/plasma_protocol.c]: Indentation here (and other places in this comment)

File: src/plasma/plasma.h [hunk 2]
```diff
@@ -16,6 +16,8 @@
 #include "utarray.h"
 #include "uthash.h"
 
+#define IPv4_ADDR_LEN 4
+
 /**
```
Comment (robertnishihara) [inline src/plasma/plasma.h]: A bunch of stuff in this file can go away, right?

File: src/plasma/plasma.h [hunk 3]
```diff
@@ -16,6 +16,8 @@
 #include "utarray.h"
 #include "uthash.h"
 
+#define IPv4_ADDR_LEN 4
```
Comment (robertnishihara) [inline src/plasma/plasma.h]: We aren't using this anywhere, right?

File: src/plasma/plasma_manager.c [hunk 4]
```diff
@@ -1234,36 +1226,53 @@ void process_message(event_loop *loop,
                      int events) {
   client_connection *conn = (client_connection *) context;
 
+  int64_t length;
   int64_t type;
-  plasma_request *req;
-  CHECK(plasma_receive_request(client_sock, &type, &req) >= 0);
+  uint8_t *data;
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.c]: where does this get freed?

File: src/plasma/plasma_protocol.c [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+}
+
+/**
+ * Finalize the flatbuffers and write a message with the result to a
+ * file descriptor.
+ *
+ * @param B Pointer to the flatbuffer builder.
+ * @param fd File descriptor the message gets written to.
+ * @param message_type Type of the message that is written.
+ *
```
Comment (robertnishihara) [inline src/plasma/plasma_protocol.c]: remove newline

File: src/plasma/plasma_client.h [hunk 6]
```diff
@@ -202,6 +202,11 @@ int64_t plasma_evict(plasma_connection *conn, int64_t num_bytes);
  *
  * @param conn The object containing the connection state.
  * @param num_object_ids The number of object IDs fetch is being called on.
+ * This method is idempotent in the sense that it is ok to call it multiple
+ * times.
+ *
+ * @param conn The object containing the connection state.
+ * @param num_object_ids The number of object IDs fetch is being called on.
```
Comment (robertnishihara) [inline src/plasma/plasma_client.h]: These 5 lines were added accidentally, right?

File: src/plasma/plasma_manager.c [hunk 7]
```diff
@@ -193,7 +194,7 @@ struct plasma_manager_state {
   client_connection *manager_connections;
   db_handle *db;
   /** Our address. */
-  uint8_t addr[4];
+  const char *addr;
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.c]: When you rebase, you will need to update the check I added in https://github.com/ray-project/ray/blob/master/src/plasma/plasma_manager.c#L836

File: src/plasma/plasma_manager.c [hunk 8]
```diff
@@ -563,23 +569,17 @@ void send_queued_request(event_loop *loop,
...
   case PLASMA_DATA:
     LOG_DEBUG("Transferring object to manager");
     if (conn->cursor == 0) {
       /* If the cursor is zero, we haven't sent any requests for this object
        * yet,
        * so send the initial PLASMA_DATA request. */
-      manager_req.data_size = buf->data_size;
-      manager_req.metadata_size = buf->metadata_size;
-      CHECK(plasma_send_request(conn->fd, PLASMA_DATA, &manager_req) >= 0);
+      CHECK(plasma_send_DataReply(conn->fd, state->builder, buf->object_id, buf->data_size, buf->metadata_size) >= 0);
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.c]: Why is this a DataReply, shouldn't it be a DataRequest?

File: src/plasma/plasma_manager.c [hunk 9]
```diff
@@ -563,23 +569,17 @@ void send_queued_request(event_loop *loop,
...
 
   plasma_request_buffer *buf = conn->transfer_queue;
-  plasma_request manager_req = plasma_make_request(buf->object_id);
   switch (buf->type) {
   case PLASMA_TRANSFER:
-    memcpy(manager_req.addr, conn->manager_state->addr,
-           sizeof(manager_req.addr));
-    manager_req.port = conn->manager_state->port;
-    CHECK(plasma_send_request(conn->fd, PLASMA_TRANSFER, &manager_req) >= 0);
+    CHECK(plasma_send_DataRequest(conn->fd, state->builder, buf->object_id, state->addr, state->port) >= 0);
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.c]: Why is this a DataRequest, shouldn't it be a TransferRequest? Or did you change the naming convention here?

File: src/plasma/plasma_store.c [hunk 10]
```diff
@@ -333,20 +337,20 @@ void seal_object(client *client_context,
...
+    object.data_size = entry->info.data_size;
+    object.metadata_size = entry->info.metadata_size;
     HASH_DELETE(handle, plasma_state->objects_notify, notify_entry);
     /* Send notifications to the clients that were waiting for this object. */
     for (int i = 0; i < utarray_len(notify_entry->waiting_clients); ++i) {
       client **c = (client **) utarray_eltptr(notify_entry->waiting_clients, i);
-      CHECK(plasma_send_reply((*c)->sock, &reply) >= 0);
-      CHECK(send_fd((*c)->sock, reply.object.handle.store_fd) >= 0);
+      CHECK(plasma_send_GetReply((*c)->sock, plasma_state->builder, &object_id, &object, 1) >= 0);
+      // CHECK(plasma_send_reply((*c)->sock, &reply) >= 0);
```
Comment (robertnishihara) [inline src/plasma/plasma_store.c]: Remove this

File: src/plasma/test/serialization_tests.c [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+SUITE(plasma_serialization_tests);
+
+protocol_builder *g_B;
+
+/**
+ * Create a temporary file. Needs to be closed by the caller.
+ *
+ * @return File descriptor of the file.
+ */
+int create_temp_file() {
```
Comment (robertnishihara) [inline src/plasma/test/serialization_tests.c]: probably need to make it `int create_temp_file(void) {`

File: src/plasma/test/serialization_tests.c [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+  lseek(fd, 0, SEEK_SET);
+  int64_t type;
+  int64_t length;
+  uint8_t *data;
+  read_message(fd, PLASMA_PROTOCOL_VERSION, &type, &length, &data);
+  CHECK(type == message_type);
+  return data;
+}
+
+plasma_object random_plasma_object() {
```
Comment (robertnishihara) [inline src/plasma/test/serialization_tests.c]: may need to pass in `void`

File: src/plasma/plasma_protocol.c [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+  return reply_data;
+}
+
+int plasma_send_CreateRequest(int sock,
+                              protocol_builder *B,
+                              object_id object_id,
+                              int64_t data_size,
+                              int64_t metadata_size) {
+  PlasmaCreateRequest_start_as_root(B);
+  PlasmaCreateRequest_object_id_create(B, (const char *)&object_id.id[0], UNIQUE_ID_SIZE);
```
Comment (robertnishihara) [inline src/plasma/plasma_protocol.c]: probably need a space after the cast

File: src/plasma/plasma_client.c [hunk 14]
```diff
@@ -250,13 +253,18 @@ void plasma_get(plasma_connection *conn,
...
-    plasma_reply reply;
-    CHECK(plasma_receive_reply(conn->store_conn, sizeof(reply), &reply) >= 0);
+    CHECK(plasma_send_GetRequest(conn->store_conn, conn->builder, &obj_id, 1) >= 0);
+    uint8_t *reply_data = plasma_receive(conn->store_conn, MessageType_PlasmaGetReply);
+    object_id* gotten_object_id;
+    int64_t num_objects;
+    plasma_read_GetReply(reply_data, &gotten_object_id, &object_data, &num_objects);
+    free(reply_data);
+    DCHECK(num_objects == 1);
+    DCHECK(memcmp(gotten_object_id, &obj_id, UNIQUE_ID_SIZE) == 0);
```
Comment (robertnishihara) [inline src/plasma/plasma_client.c]: Can we replace all instances of `UNIQUE_ID_SIZE` with `sizeof(obj_id)`, or whatever the appropriate thing is?

**Comment by pcmoritz** [comment]
merged in https://github.com/ray-project/ray/pull/140

**Pull request closed without merging**